src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java

```diff
@@ -85,11 +99,61 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
     protected abstract String parserName();
 
     protected abstract void doXContent(XContentBuilder builder, Params params) throws IOException;
-
     @Override
     public QueryValidationException validate() {
         // default impl does not validate, subclasses should override.
         //norelease to be removed once all queries support validation
         return null;
     }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        //norelease all querybuilders should finally implement their own streaming
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+      //norelease all querybuilders should finally implement their own streaming
+    }
+
+    protected static QueryBuilder readQueryBuilderFrom(StreamInput in) throws IOException {
+        QueryBuilder newBuilder;
+        String name = in.readString();
+        Class<?> newBuilderClass = queryBuilderRegistry.get(name);
+        if (newBuilderClass != null) {
+            try {
+                newBuilder = (QueryBuilder) newBuilderClass.newInstance();
```

Comment:
Wonder if this can be moved to StreamInput somehow, but not sure where the registry would go then....